Duplicate Structure Names causes Inability to Compile Composed Reactors for Arduino

The following example for Arduino is unable to compile with the arduino-cli toolchain. I am using lf-cli-0.5.2-SNAPSHOT-nightly-20231015050236-Linux-x86_64; this issue exists on lf-cli-0.5.1-Linux-x86_64 as well. I am currently testing on Windows Pro 11 WSL with an Ubuntu 22.04 VM.


dualreactorblink.lf
```
target C {
    platform: {
        name: "arduino",
        board: "arduino:avr:uno"
    }
}

reactor Blinker {
    input in_port:bool

    reaction(startup) {=
        pinMode(LED_BUILTIN, OUTPUT);
    =}

    reaction(in_port) {=
        digitalWrite(LED_BUILTIN, in_port->value ? HIGH : LOW);
    =}
}

reactor Timer {
    timer t1(0, 500 msec)
    state on_off:bool = false
    output out_port:bool

    reaction (t1) -> out_port {=
        self->on_off = !self->on_off;
        lf_set(out_port, self->on_off);
    =}
}

main reactor {
    the_blinker = new Blinker()
    the_timer = new Timer()
    the_timer.out_port -> the_blinker.in_port;
}
```


I get the following errors when I run lfc:
In file included from /xxx/lf/src-gen/dualreactorblink/_dualreactorblink_main.ino:2:0:
/xxx/lf/src-gen/dualreactorblink/include/dualreactorblink/dualreactorblink.h:26:3: error: conflicting declaration 'typedef struct blinker_in_port_t blinker_in_port_t'
 } blinker_in_port_t;
   ^~~~~~~~~~~~~~~~~
In file included from /xxx/lf/src-gen/dualreactorblink/_blinker.ino:2:0:
/xxx/lf/src-gen/dualreactorblink/include/dualreactorblink/Blinker.h:31:3: note: previous declaration as 'typedef struct blinker_in_port_t blinker_in_port_t'
 } blinker_in_port_t;
   ^~~~~~~~~~~~~~~~~
In file included from /xxx/lf/src-gen/dualreactorblink/_timer.ino:2:0:
/xxx/lf/src-gen/dualreactorblink/include/dualreactorblink/Timer.h:32:3: error: conflicting declaration 'typedef struct timer_out_port_t timer_out_port_t'
 } timer_out_port_t;
   ^~~~~~~~~~~~~~~~
In file included from /xxx/lf/src-gen/dualreactorblink/_dualreactorblink_main.ino:2:0:
/xxx/lf/src-gen/dualreactorblink/include/dualreactorblink/dualreactorblink.h:35:3: note: previous declaration as 'typedef struct timer_out_port_t timer_out_port_t'
 } timer_out_port_t;
   ^~~~~~~~~~~~~~~~
   
Looking through the generated files, I observed that struct blinker_in_port_t is declared in both Blinker.h and dualreactorblink.h. The definitions are different, though. The structure in Blinker.h includes some FEDERATED ifdefs. Other than that, the two structures are the same. Tracing through the .h and .ino files, both of the header files end up included for the build, which looks like it causes the conflicting declaration error.

I do not know if this is an error with how I have set up the program, but I followed the instructions here (https://www.lf-lang.org/docs/handbook/arduino?target=c) to set up the arduino compile and the instructions here (https://www.lf-lang.org/docs/handbook/composing-reactors?target=c) to make the example. 

Thank you. Please let me know if there is additional information I can add to help clarify this.

--- a/core/src/main/java/org/lflang/generator/c/CGenerator.java
+++ b/core/src/main/java/org/lflang/generator/c/CGenerator.java
@@ -948,7 +948,7 @@ public class CGenerator extends GeneratorBase {
     CodeBuilder header = new CodeBuilder();
     CodeBuilder src = new CodeBuilder();
     final String headerName = CUtil.getName(tpr) + ".h";
-    var guardMacro = headerName.toUpperCase().replace(".", "_");
+    var guardMacro = CUtil.internalIncludeGuard(tpr);
     header.pr("#ifndef " + guardMacro);
     header.pr("#define " + guardMacro);
     generateReactorClassHeaders(tpr, headerName, header, src);

--- a/core/src/main/java/org/lflang/generator/c/CReactorHeaderFileGenerator.java
+++ b/core/src/main/java/org/lflang/generator/c/CReactorHeaderFileGenerator.java
@@ -55,7 +55,7 @@ public class CReactorHeaderFileGenerator {
       GenerateAuxiliaryStructs generator,
       String topLevelPreamble) {
     CodeBuilder builder = new CodeBuilder();
-    appendIncludeGuard(builder, tpr);
+    appendIncludeGuards(builder, tpr);
     builder.pr(topLevelPreamble);
     appendPoundIncludes(builder);
     tpr.doDefines(builder);
@@ -64,14 +64,20 @@ public class CReactorHeaderFileGenerator {
     for (Reaction reaction : tpr.reactor().getReactions()) {
       appendSignature(builder, types, reaction, tpr);
     }
-    builder.pr("#endif");
+    closeIncludeGuards(builder);
     return builder.getCode();
   }
 
-  private static void appendIncludeGuard(CodeBuilder builder, TypeParameterizedReactor r) {
+  private static void appendIncludeGuards(CodeBuilder builder, TypeParameterizedReactor r) {
     String macro = CUtil.getName(r) + "_H";
     builder.pr("#ifndef " + macro);
     builder.pr("#define " + macro);
+    builder.pr("#ifndef " + CUtil.internalIncludeGuard(r));
+  }
+
+  private static void closeIncludeGuards(CodeBuilder builder) {
+    builder.pr("#endif");
+    builder.pr("#endif");
   }
 
   private static void appendPoundIncludes(CodeBuilder builder) {

--- a/core/src/main/java/org/lflang/generator/c/CReactorHeaderFileGenerator.java
+++ b/core/src/main/java/org/lflang/generator/c/CReactorHeaderFileGenerator.java
@@ -72,7 +72,11 @@ public class CReactorHeaderFileGenerator {
     String macro = CUtil.getName(r) + "_H";
     builder.pr("#ifndef " + macro);
     builder.pr("#define " + macro);
-    builder.pr("#ifndef " + CUtil.internalIncludeGuard(r));
+    builder.pr(
+        "#ifndef "
+            + CUtil.internalIncludeGuard(r)
+            + " // necessary for arduino-cli, which automatically includes headers that are not"
+            + " used");
   }
 
   private static void closeIncludeGuards(CodeBuilder builder) {

--- a/core/src/main/java/org/lflang/generator/c/CUtil.java
+++ b/core/src/main/java/org/lflang/generator/c/CUtil.java
@@ -148,6 +148,12 @@ public class CUtil {
     return name;
   }
 
+  /** Return the name used in the internal (non-user-facing) include guard for the given reactor. */
+  public static String internalIncludeGuard(TypeParameterizedReactor tpr) {
+    final String headerName = CUtil.getName(tpr) + ".h";
+    return headerName.toUpperCase().replace(".", "_");
+  }
+
   /**
    * Return a reference to the specified port.
    *

--- a/test/C/src/arduino/DualReactorBlink.lf
+++ b/test/C/src/arduino/DualReactorBlink.lf
@@ -1,0 +1,35 @@
+target C {
+  platform: {
+    name: "arduino",
+    board: "arduino:avr:mega"
+  }
+}
+
+reactor Blinker {
+  input in_port: bool
+
+  reaction(startup) {=
+    pinMode(LED_BUILTIN, OUTPUT);
+  =}
+
+  reaction(in_port) {=
+    digitalWrite(LED_BUILTIN, in_port->value ? HIGH : LOW);
+  =}
+}
+
+reactor Timer {
+  timer t1(0, 500 msec)
+  state on_off: bool = false
+  output out_port: bool
+
+  reaction(t1) -> out_port {=
+    self->on_off = !self->on_off;
+    lf_set(out_port, self->on_off);
+  =}
+}
+
+main reactor {
+  the_blinker = new Blinker()
+  the_timer = new Timer()
+  the_timer.out_port -> the_blinker.in_port
+}

